src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,265 @@
+{
+    return !ldap_server.empty() &&/* !roles.empty() &&*/ access_control_manager;
+}
+
+
+void LDAPAccessStorage::processRoleChange(const UUID & id, const AccessEntityPtr & entity)
+{
+    auto role_ptr = typeid_cast<std::shared_ptr<const Role>>(entity);
+    if (role_ptr)
+    {
+        if (default_role_names.find(role_ptr->getName()) != default_role_names.end())
+        {
+            auto update_func = [&id](const AccessEntityPtr & cached_entity) -> AccessEntityPtr
+            {
+                auto user_ptr = typeid_cast<std::shared_ptr<const User>>(cached_entity);
+                if (user_ptr && user_ptr->granted_roles.roles.find(id) == user_ptr->granted_roles.roles.end())
+                {
+                    auto clone = user_ptr->clone();
+                    auto user_clone_ptr = typeid_cast<std::shared_ptr<User>>(clone);
+                    user_clone_ptr->granted_roles.grant(id);
+                    return user_clone_ptr;
+                }
+                return cached_entity;
+            };
+
+            memory_storage.update(memory_storage.findAll<User>(), update_func);
+            roles_of_interest.insert(id);
+        }
+    }
+    else
+    {
+        if (roles_of_interest.find(id) != roles_of_interest.end())
+        {
+            auto update_func = [&id](const AccessEntityPtr & cached_entity) -> AccessEntityPtr
+            {
+                auto user_ptr = typeid_cast<std::shared_ptr<const User>>(cached_entity);
+                if (user_ptr && user_ptr->granted_roles.roles.find(id) != user_ptr->granted_roles.roles.end())
+                {
+                    auto clone = user_ptr->clone();
+                    auto user_clone_ptr = typeid_cast<std::shared_ptr<User>>(clone);
+                    user_clone_ptr->granted_roles.revoke(id);
+                    return user_clone_ptr;
+                }
+                return cached_entity;
+            };
+
+            memory_storage.update(memory_storage.findAll<User>(), update_func);
+            roles_of_interest.erase(id);
+        }
+    }
+}
+
+
+const char * LDAPAccessStorage::getStorageType() const
+{
+    return STORAGE_TYPE;
+}
+
+
+bool LDAPAccessStorage::isStorageReadOnly() const
+{
+    return true;
+}
+
+
+std::optional<UUID> LDAPAccessStorage::findImpl(EntityType type, const String & name) const
+{
+    return memory_storage.find(type, name);
+}
+
+
+std::optional<UUID> LDAPAccessStorage::findOrGenerateImpl(EntityType type, const String & name) const
+{
+    if (type == EntityType::USER)
+    {
+        std::scoped_lock lock(mutex);
+
+        // Return the id immediately if we already have it.
+        const auto id = memory_storage.find(type, name);
+        if (id.has_value())
+            return id;
+
+        if (!isConfiguredNoLock())
+            return {};
+
+        // Stop if entity exists anywhere else, to avoid generating duplicates.
+        const auto * this_base = dynamic_cast<const IAccessStorage *>(this);
```

Comment:
With the new `IAccessStorage::login()` function that code becomes excessive.

Reply:
Removed.